git/clone.go

```diff
@@ -132,5 +143,30 @@ func Clone(opts CloneOptions, rmt Remote, dst File) error {
 	// Finally, checkout the files. Since it's an initial clone, we just
 	// do a hard reset and don't try to be intelligent about what readtree
 	// does.
+	//
+	// We need to be sure we're within the repo, so that ReadTree and
+	// CheckoutIndexUncommitted don't have problems making paths relative,
+	// but then we restore the environment and c when we're done.
+	gwd := c.WorkDir
+	ggd := c.GitDir
+	pwd, err := os.Getwd()
+	defer func() {
+		os.Chdir(pwd)
+		c.WorkDir = gwd
+		c.GitDir = ggd
+	}()
+	if err != nil {
+		return err
+	}
+	if err := os.Chdir(c.WorkDir.String()); err != nil {
+		return err
+	}
+	// filepath.Rel can if workdir is ".", so use the absolute path
+	absdir, err := filepath.Abs(".")
+	if err != nil {
+		return err
+	}
+	c.WorkDir = WorkDir(absdir)
+	c.GitDir = GitDir(filepath.Join(c.WorkDir.String(), ".git"))
```

Comment:
I wonder if this would be a problem if there was a bare repo being cloned.

Reply:
I don't think it would make a difference with the repo being cloned, I think it might break if --bare is passed to clone, but I doubt that's ever worked

Reply:
was easy enough to fix by just shortcircuiting the reset if --bare is specified..